mp/src/game/shared/momentum/mom_stickybomb.cpp

```diff
@@ -0,0 +1,387 @@
+
+#define MOM_STICKYBOMB_MODEL "models/weapons/w_models/w_stickybomb.mdl"
+#define MOM_STICKYBOMB_RADIUS 146.0f
+#define MOM_STICKYBOMB_INITIAL_SPEED 900.0f
+#define MOM_STICKYBOMB_MAX_SPEED 2400.0f
+#define MOM_STICKYBOMB_GRAVITY 0.5f
+#define MOM_STICKYBOMB_FRICTION 0.8f
+#define MOM_STICKYBOMB_ELASTICITY 0.45f
+
+#ifndef CLIENT_DLL
+BEGIN_DATADESC(CMomStickybomb)
+    // Fields
+    DEFINE_FIELD(m_hLauncher, FIELD_EHANDLE),
+    DEFINE_FIELD(m_flDamage, FIELD_FLOAT),
+    DEFINE_FIELD(m_bTouched, FIELD_BOOLEAN),
+    DEFINE_FIELD(m_bPulsed, FIELD_BOOLEAN),
+
+    // Functions
+    DEFINE_ENTITYFUNC(Touch),
+END_DATADESC();
+#endif
+
+IMPLEMENT_NETWORKCLASS_ALIASED(MomStickybomb, DT_MomStickybomb)
+
+BEGIN_NETWORK_TABLE(CMomStickybomb, DT_MomStickybomb)
+#ifdef CLIENT_DLL
+RecvPropVector(RECVINFO(m_vInitialVelocity)),
+RecvPropInt(RECVINFO(m_bTouched)),
+RecvPropVector(RECVINFO_NAME(m_vecNetworkOrigin, m_vecOrigin)),
+RecvPropEHandle(RECVINFO(m_hLauncher)),
+#else
+SendPropVector(SENDINFO(m_vInitialVelocity),
+               20,    // nbits
+               0,     // flags
+               -3000, // low value
+               3000   // high value
+               ),
+SendPropExclude("DT_BaseEntity", "m_vecOrigin"),
+SendPropBool(SENDINFO(m_bTouched)),
+SendPropVector(SENDINFO(m_vecOrigin), -1, SPROP_COORD_MP_INTEGRAL | SPROP_CHANGES_OFTEN, 0.0f, HIGH_DEFAULT, SendProxy_Origin),
+SendPropEHandle(SENDINFO(m_hLauncher)),
+#endif
+END_NETWORK_TABLE();
+
+LINK_ENTITY_TO_CLASS(momentum_stickybomb, CMomStickybomb);
+PRECACHE_WEAPON_REGISTER(momentum_stickybomb);
+
+#ifdef CLIENT_DLL
+static MAKE_CONVAR(mom_sj_stickybomb_drawdelay, "0.1", FCVAR_ARCHIVE | FCVAR_CLIENTDLL,
+                   "Determines how long it takes for stickies to start being drawn upon spawning.\n", 0, 1);
+#endif
+
+CMomStickybomb::CMomStickybomb()
+{
+    m_vInitialVelocity.Init();
+    m_bTouched = false;
+    m_flChargeTime = 0.0f;
+#ifdef CLIENT_DLL
+    m_flSpawnTime = 0.0f;
+#else
+    m_flDamage = 0.0f;
+    m_flCreationTime = 0.0f;
+    m_bUseImpactNormal = false;
+    m_vecImpactNormal.Init();
+#endif
+}
+
+CMomStickybomb::~CMomStickybomb()
+{
+#ifdef CLIENT_DLL
+    ParticleProp()->StopEmission();
+#endif
+}
+
+void CMomStickybomb::UpdateOnRemove()
+{
+    // Tell our launcher that we were removed
+    CMomentumStickybombLauncher *pLauncher = dynamic_cast<CMomentumStickybombLauncher *>(m_hLauncher.Get());
+
+    if (pLauncher)
+    {
+        pLauncher->DeathNotice(this);
+    }
+
+    BaseClass::UpdateOnRemove();
+}
+
+#ifdef CLIENT_DLL
+
+void CMomStickybomb::OnDataChanged(DataUpdateType_t type)
+{
+    BaseClass::OnDataChanged(type);
+
+    if (type == DATA_UPDATE_CREATED)
+    {
+        m_flCreationTime = gpGlobals->curtime;
+        ParticleProp()->Create("stickybombtrail_red", PATTACH_ABSORIGIN_FOLLOW);
+        m_bPulsed = false;
+
+        CMomentumStickybombLauncher *pLauncher = dynamic_cast<CMomentumStickybombLauncher *>(m_hLauncher.Get());
```

Comment:
`m_hLauncher` should be of type `CMomentumStickybombLauncher`, so there's no need for the dynamic cast.